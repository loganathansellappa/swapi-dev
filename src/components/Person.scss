.personInfo {
  background-image: linear-gradient(260deg, #2376ae 0%, #c16ecf 100%);
  border-radius: 0.5rem;
  overflow: hidden;
  margin: 1%;
  display: flex;
  color: #ffffff;
  flex-basis: 400px;
  flex-direction: column;
  .favourite {
    cursor: pointer;
    display: flex;
    flex-direction: column;
    align-items: center;
    row-gap: 10px;
  }

  .name {
    text-align: center;
    color: #fff;
    text-decoration: none;
  }

  .name:hover,
  .favourite:hover {
    animation: shake 0.82s cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
    color: yellowgreen;
    transform: translate3d(0, 0, 0);
    backface-visibility: hidden;
    perspective: 1000px;
  }
  .favouriteIcon {
    font-size: 50px;
  }

  .propContainer {
    display: flex;
    justify-content: center;
    .props {
      display: flex;
      justify-content: space-between;
      padding: 10px;
      gap: 5px;
    }
  }

  @keyframes shake {
    10%,
    90% {
      transform: translate3d(-1px, 0, 0);
    }

    20%,
    80% {
      transform: translate3d(2px, 0, 0);
    }

    30%,
    50%,
    70% {
      transform: translate3d(-4px, 0, 0);
    }

    40%,
    60% {
      transform: translate3d(4px, 0, 0);
    }
  }
}
