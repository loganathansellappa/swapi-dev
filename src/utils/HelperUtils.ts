export const apiService = () => `${process.env.SWAP_URL}`;
export const capitalize = (s = 'n/a') => s[0].toUpperCase() + s.slice(1);
