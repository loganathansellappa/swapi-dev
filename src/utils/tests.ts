import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';

export const renderWithRouter = (ui: JSX.Element, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: BrowserRouter }),
  };
};
export const testStore = () => {
  const mockStore = configureStore();
  const initialState = { favoritePeople: { favourites: [] } };
  const store = mockStore(initialState);
  return store;
};

export const mockResource = () => {
  return {
    name: 'Luke Skywalker',
    height: '172',
    mass: '77',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    birth_year: '19BBY',
    gender: 'male',
    homeworld: 'https://swapi.dev/api/planets/1/',
    films: [
      'https://swapi.dev/api/films/1/',
      'https://swapi.dev/api/films/2/',
      'https://swapi.dev/api/films/3/',
      'https://swapi.dev/api/films/6/',
    ],
    species: [],
    vehicles: [
      'https://swapi.dev/api/vehicles/14/',
      'https://swapi.dev/api/vehicles/30/',
    ],
    starships: [
      'https://swapi.dev/api/starships/12/',
      'https://swapi.dev/api/starships/22/',
    ],
    created: '2014-12-09T13:50:51.644000Z',
    edited: '2014-12-20T21:17:56.891000Z',
    url: 'https://swapi.dev/api/people/1/',
  };
};
