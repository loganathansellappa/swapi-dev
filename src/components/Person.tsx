import React from 'react';
import { AnyResourceType, ResourceType } from '../@types/People';
import { RootState } from '../store';
import { useDispatch, useSelector } from 'react-redux';
import { add, remove } from '../states/People';
import { Link } from 'react-router-dom';
import { faClover, faHeart } from '@fortawesome/free-solid-svg-icons';
import { capitalize } from '../utils/HelperUtils';
import './Person.scss';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface PersonProps {
  person: ResourceType;
}

export const Person = ({ person }: PersonProps) => {
  const dispatch = useDispatch();
  const isFavourite = useSelector((state: RootState) =>
    state.favoritePeople.favourites?.find((p) => p.url === person.url)
  );

  const resourceId = person.url?.match(/\d/g).join('');
  const setFavourite = () => {
    if (isFavourite) {
      dispatch(remove(person.url));
      return;
    }
    dispatch(add(person));
  };

  const typeInfo = (data: AnyResourceType) => {
    if (data.episode_id) {
      return 'Film';
    }
    if (data.rotation_period) {
      return 'Planet';
    }
    if (data.manufacturer) {
      return 'StarShip';
    }
    if (data.vehicle_class) {
      return 'Vehicle';
    }
    return 'People';
  };

  return (
    <div className="personInfo">
      <Link className={'name'} to={`resource/people/${resourceId}`}>
        <h3>{person.name}</h3>
        <p>{typeInfo(person)}</p>
      </Link>

      <div className="propContainer">
        <div className="props">
          <div className="label">Gender</div>
          <div className="value">{capitalize(person.gender)}</div>
        </div>
        <div className="props">
          <div className="label">Height</div>
          <div className="value">{person.height || 'N/A'}</div>
        </div>
      </div>
      <div className="favourite" onClick={setFavourite}>
        <div>Mark favourite</div>
        <div>
          <FontAwesomeIcon
            className={'favouriteIcon'}
            icon={faHeart}
            color={isFavourite ? 'red' : 'white'}
          />
        </div>
      </div>
    </div>
  );
};
