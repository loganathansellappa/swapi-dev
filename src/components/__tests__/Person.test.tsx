import { mockResource, renderWithRouter, testStore } from '../..//utils/tests';
import { Person } from '../Person';
import { Provider } from 'react-redux';

describe('Person', () => {
  let element: { container: HTMLElement };
  beforeEach(() => {
    element = renderWithRouter(
      <Provider store={testStore()}>
        <Person person={mockResource()} />
      </Provider>
    );
  });

  test('renders Person component', () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });
});
