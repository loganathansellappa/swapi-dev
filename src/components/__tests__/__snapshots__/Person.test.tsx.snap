// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person renders Person component 1`] = `
<div
  class="personInfo"
>
  <a
    class="name"
    href="/resource/people/1"
  >
    <h3>
      Luke Skywalker
    </h3>
    <p>
      People
    </p>
  </a>
  <div
    class="propContainer"
  >
    <div
      class="props"
    >
      <div
        class="label"
      >
        Gender
      </div>
      <div
        class="value"
      >
        Male
      </div>
    </div>
    <div
      class="props"
    >
      <div
        class="label"
      >
        Height
      </div>
      <div
        class="value"
      >
        172
      </div>
    </div>
  </div>
  <div
    class="favourite"
  >
    <div>
      Mark favourite
    </div>
    <div>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-heart favouriteIcon"
        color="white"
        data-icon="heart"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 190.9V185.1C0 115.2 50.52 55.58 119.4 44.1C164.1 36.51 211.4 51.37 244 84.02L256 96L267.1 84.02C300.6 51.37 347 36.51 392.6 44.1C461.5 55.58 512 115.2 512 185.1V190.9C512 232.4 494.8 272.1 464.4 300.4L283.7 469.1C276.2 476.1 266.3 480 256 480C245.7 480 235.8 476.1 228.3 469.1L47.59 300.4C17.23 272.1 .0003 232.4 .0003 190.9L0 190.9z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
